File: mobile/test/telemetry_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:mytube_mobile/telemetry.dart';

class MemoryTelemetryStorage implements TelemetryStorage {
  final values = <String, String>{};
  @override
  Future<void> delete(String key) async => values.remove(key);
  @override
  Future<String?> read(String key) async => values[key];
  @override
  Future<void> write(String key, String value) async => values[key] = value;
}

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  test('queues offline and removes accepted events', () async {
    var calls = 0;
    final client = MockClient((request) async {
      calls++;
      if (calls == 1) throw http.ClientException('offline');
      return http.Response('{}', 200);
    });
    var sequence = 0;
    final telemetry = MobileTelemetry(
      storage: MemoryTelemetryStorage(),
      client: client,
      randomId: () => 'event_identifier_${++sequence}',
      startBackground: false,
    );
    await telemetry.initialize(
      baseUrl: 'https://api.example.test',
      token: 'token',
      appVersion: '1.0.0',
    );
    await telemetry.flush(ignoreBackoff: true);
    expect(telemetry.queuedCount, 1);
    await telemetry.flush(ignoreBackoff: true);
    expect(telemetry.queuedCount, 0);
    await telemetry.dispose();
  });

  test('disabled analytics clears queued events', () async {
    final telemetry = MobileTelemetry(
      storage: MemoryTelemetryStorage(),
      client: MockClient((_) async => http.Response('{}', 200)),
      startBackground: false,
    );
    await telemetry.initialize(
      baseUrl: 'https://api.example.test',
      token: '',
      appVersion: '1.0.0',
    );
    expect(telemetry.queuedCount, 1);
    await telemetry.setEnabled(false);
    expect(telemetry.queuedCount, 0);
    telemetry.track(TelemetryEventName.videoStarted);
    expect(telemetry.queuedCount, 0);
    await telemetry.dispose();
  });

  test('payload excludes video metadata and token', () async {
    String body = '';
    final telemetry = MobileTelemetry(
      storage: MemoryTelemetryStorage(),
      client: MockClient((request) async {
        body = request.body;
        return http.Response('{}', 200);
      }),
      startBackground: false,
    );
    await telemetry.initialize(
      baseUrl: 'https://api.example.test',
      token: 'secret-token',
      appVersion: '1.0.0',
    );
    telemetry.track(
      TelemetryEventName.playbackFailed,
      playbackMode: 'standalone',
      outcomeCode: 'network',
    );
    await telemetry.flush(ignoreBackoff: true);
    final decoded = jsonDecode(body) as Map<String, dynamic>;
    expect(decoded['schema_version'], 1);
    expect(body, isNot(contains('secret-token')));
    expect(body, isNot(contains('title')));
    expect(body, isNot(contains('url')));
    await telemetry.dispose();
  });
}

File: frontend/src/telemetry.test.ts
import { describe, expect, it, vi } from 'vitest';
import { TelemetryClient } from './telemetry';

function memoryStorage() {
  const values = new Map<string, string>();
  return {
    getItem: (key: string) => values.get(key) ?? null,
    setItem: (key: string, value: string) => { values.set(key, value); },
    removeItem: (key: string) => { values.delete(key); },
  };
}

function createClient(fetchImpl: typeof fetch, now = Date.now()) {
  let sequence = 0;
  const storage = memoryStorage();
  const sessionStorage = memoryStorage();
  const client = new TelemetryClient({
    storage,
    sessionStorage,
    fetch: fetchImpl,
    now: () => now,
    randomID: () => `event_identifier_${++sequence}`,
    apiBase: () => 'https://api.example.test',
    appVersion: () => '1.0.0',
  });
  return { client, storage, sessionStorage };
}

describe('telemetry client', () => {
  it('queues offline and removes an accepted batch exactly once', async () => {
    const fetchImpl = vi.fn<typeof fetch>()
      .mockRejectedValueOnce(new Error('offline'))
      .mockResolvedValueOnce(new Response('{}', { status: 200 }));
    const { client } = createClient(fetchImpl);
    client.track('video_started', { playback_mode: 'standalone' });

    await client.flush();
    expect(client.queuedCount()).toBe(1);
    await client.flush(true);
    expect(client.queuedCount()).toBe(0);
    expect(fetchImpl).toHaveBeenCalledTimes(2);
  });

  it('clears queued data when analytics is disabled', () => {
    const { client } = createClient(vi.fn<typeof fetch>());
    client.track('app_opened');
    expect(client.queuedCount()).toBe(1);
    client.setEnabled(false);
    expect(client.queuedCount()).toBe(0);
    client.track('video_started');
    expect(client.queuedCount()).toBe(0);
  });

  it('does not include sensitive properties in the encoded event', async () => {
    const fetchImpl = vi.fn<typeof fetch>().mockResolvedValue(new Response('{}', { status: 200 }));
    const { client } = createClient(fetchImpl);
    client.track('playback_failed', { outcome_code: 'network' });
    await client.flush();

    const init = fetchImpl.mock.calls[0]?.[1];
    const body = String(init?.body);
    expect(body).toContain('outcome_code');
    expect(body).not.toContain('title');
    expect(body).not.toContain('url');
    expect(init?.headers).not.toHaveProperty('Authorization');
  });

  it('records app opened once per tab session', () => {
    const { client } = createClient(vi.fn<typeof fetch>());
    client.trackOpenedOnce();
    client.trackOpenedOnce();
    expect(client.queuedCount()).toBe(1);
  });
});
